File: scripts/generate_insights.py
"""Генерация AI-инсайтов по данным заказов через OpenAI API."""

import os
from collections import defaultdict

from dotenv import load_dotenv
from openai import OpenAI
from supabase import create_client, Client

# --- Константы ---

load_dotenv()

SUPABASE_URL: str = os.getenv("SUPABASE_URL", "")
SUPABASE_KEY: str = os.getenv("SUPABASE_KEY", "")
OPENAI_API_KEY: str = os.getenv("OPENAI_API_KEY", "")
MODEL: str = "gpt-4o-mini"


# --- Функции ---


def fetch_data(supabase: Client) -> tuple[list[dict], list[dict]]:
    """Загружает заказы и товары из Supabase.

    Args:
        supabase: Клиент Supabase.

    Returns:
        Кортеж (список заказов, список товаров).
    """
    orders = supabase.table("orders").select("*").execute().data or []
    items = supabase.table("order_items").select("*").execute().data or []
    return orders, items


def build_summary(orders: list[dict], items: list[dict]) -> str:
    """Формирует текстовую сводку данных для передачи в OpenAI.

    Args:
        orders: Список заказов из таблицы orders.
        items: Список товаров из таблицы order_items.

    Returns:
        Строка с аналитической сводкой.
    """
    total_count = len(orders)
    total_sum = sum(o.get("total", 0) or 0 for o in orders)
    avg_check = round(total_sum / total_count) if total_count else 0

    # Группировка по городам
    cities: dict[str, dict] = defaultdict(lambda: {"count": 0, "sum": 0})
    for o in orders:
        city = o.get("city") or "Не указан"
        cities[city]["count"] += 1
        cities[city]["sum"] += o.get("total", 0) or 0

    # Группировка по UTM-источникам
    utms: dict[str, dict] = defaultdict(lambda: {"count": 0, "sum": 0})
    for o in orders:
        utm = o.get("utm_source") or "Не указан"
        utms[utm]["count"] += 1
        utms[utm]["sum"] += o.get("total", 0) or 0

    # Группировка по товарам
    products: dict[str, dict] = defaultdict(lambda: {"qty": 0, "revenue": 0})
    for item in items:
        name = item.get("product_name") or "Неизвестно"
        products[name]["qty"] += item.get("quantity", 0)
        products[name]["revenue"] += (item.get("price", 0) or 0) * (item.get("quantity", 0) or 0)

    # Формируем текст сводки
    cities_text = "\n".join(
        f"  - {city}: {d['count']} заказов, {d['sum']:,} KZT"
        for city, d in sorted(cities.items(), key=lambda x: -x[1]["count"])
    )
    utms_text = "\n".join(
        f"  - {utm}: {d['count']} заказов, {d['sum']:,} KZT"
        for utm, d in sorted(utms.items(), key=lambda x: -x[1]["count"])
    )
    products_text = "\n".join(
        f"  - {name}: {d['qty']} шт., выручка {d['revenue']:,} KZT"
        for name, d in sorted(products.items(), key=lambda x: -x[1]["revenue"])
    )

    return f"""Данные магазина Nova (корректирующее бельё, Казахстан):

Общая статистика:
  - Заказов: {total_count}
  - Общая сумма: {total_sum:,} KZT
  - Средний чек: {avg_check:,} KZT

Города:
{cities_text}

Источники трафика (UTM):
{utms_text}

Топ товаров:
{products_text}"""


def generate_insights(summary: str, client: OpenAI) -> list[str]:
    """Отправляет сводку в OpenAI и получает список бизнес-рекомендаций.

    Args:
        summary: Текстовая сводка аналитических данных.
        client: Клиент OpenAI.

    Returns:
        Список строк — каждая строка одна рекомендация.
    """
    response = client.chat.completions.create(
        model=MODEL,
        messages=[
            {
                "role": "system",
                "content": (
                    "Ты — бизнес-аналитик интернет-магазина корректирующего белья 'Nova' (Казахстан). "
                    "Проанализируй данные заказов и дай 4-5 конкретных бизнес-рекомендаций. "
                    "Каждая рекомендация — 1-2 предложения с конкретными цифрами из данных. "
                    "Начинай каждую рекомендацию с короткого заголовка (3-5 слов), затем двоеточие и текст. "
                    "Пиши на русском. Каждая рекомендация на новой строке."
                ),
            },
            {"role": "user", "content": summary},
        ],
        temperature=0.7,
    )

    text = response.choices[0].message.content or ""
    # Разбиваем по строкам, убираем пустые
    return [line.strip() for line in text.strip().splitlines() if line.strip()]


def save_insights(supabase: Client, insights: list[str]) -> None:
    """Очищает старые инсайты и сохраняет новые в Supabase.

    Args:
        supabase: Клиент Supabase.
        insights: Список текстов рекомендаций.
    """
    supabase.table("insights").delete().neq("id", 0).execute()
    rows = [{"content": text} for text in insights]
    supabase.table("insights").insert(rows).execute()


def main() -> None:
    """Точка входа: генерирует AI-инсайты и сохраняет в Supabase."""
    if not all([SUPABASE_URL, SUPABASE_KEY, OPENAI_API_KEY]):
        print("Ошибка: заполни SUPABASE_URL, SUPABASE_KEY, OPENAI_API_KEY в .env")
        return

    supabase = create_client(SUPABASE_URL, SUPABASE_KEY)
    openai_client = OpenAI(api_key=OPENAI_API_KEY)

    print("Загружаю данные из Supabase...")
    try:
        orders, items = fetch_data(supabase)
    except Exception as e:
        print(f"Ошибка загрузки данных: {e}")
        return

    print(f"Заказов: {len(orders)}, товаров: {len(items)}")

    summary = build_summary(orders, items)

    print("Отправляю в OpenAI...")
    try:
        insights = generate_insights(summary, openai_client)
    except Exception as e:
        print(f"Ошибка OpenAI: {e}")
        return

    print("Сохраняю в Supabase...")
    try:
        save_insights(supabase, insights)
    except Exception as e:
        print(f"Ошибка сохранения: {e}")
        return

    print(f"\nСгенерировано {len(insights)} инсайтов:\n")
    for i, insight in enumerate(insights, 1):
        print(f"{i}. {insight}")


# --- Точка входа ---

if __name__ == "__main__":
    main()

File: dashboard/api/generate-insights.js
/**
 * Vercel Serverless Function: генерация AI-инсайтов.
 * Ключи читаются из переменных окружения Vercel — никогда не попадают на клиент.
 */

const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_KEY;

const SYSTEM_PROMPT =
  'Ты — бизнес-аналитик интернет-магазина корректирующего белья "Nova" (Казахстан). ' +
  'Проанализируй данные заказов и дай 4-5 конкретных бизнес-рекомендаций. ' +
  'Каждая рекомендация — 1-2 предложения с конкретными цифрами из данных. ' +
  'Начинай каждую рекомендацию с короткого заголовка (3-5 слов), затем двоеточие и текст. ' +
  'Пиши на русском. Каждая рекомендация на новой строке.';

/** Запрос к Supabase REST API */
async function supabaseFetch(path, options = {}) {
  const res = await fetch(`${SUPABASE_URL}/rest/v1/${path}`, {
    ...options,
    headers: {
      apikey: SUPABASE_SERVICE_KEY,
      Authorization: `Bearer ${SUPABASE_SERVICE_KEY}`,
      'Content-Type': 'application/json',
      Prefer: 'return=minimal',
      ...options.headers,
    },
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Supabase ${path}: ${res.status} ${text}`);
  }
  return options.method && options.method !== 'GET' ? null : res.json();
}

/** Строит текстовую сводку для промпта */
function buildSummary(orders, items) {
  const count = orders.length;
  const totalSum = orders.reduce((s, o) => s + (o.total || 0), 0);
  const avg = count ? Math.round(totalSum / count) : 0;

  const group = (arr, key) =>
    arr.reduce((acc, o) => {
      const k = o[key] || 'Не указан';
      if (!acc[k]) acc[k] = { count: 0, sum: 0 };
      acc[k].count++;
      acc[k].sum += o.total || 0;
      return acc;
    }, {});

  const fmt = (n) => n.toLocaleString('ru-RU');

  const citiesText = Object.entries(group(orders, 'city'))
    .sort((a, b) => b[1].count - a[1].count)
    .map(([c, d]) => `  - ${c}: ${d.count} заказов, ${fmt(d.sum)} KZT`)
    .join('\n');

  const utmsText = Object.entries(group(orders, 'utm_source'))
    .sort((a, b) => b[1].count - a[1].count)
    .map(([u, d]) => `  - ${u}: ${d.count} заказов, ${fmt(d.sum)} KZT`)
    .join('\n');

  const products = items.reduce((acc, i) => {
    const n = i.product_name || 'Неизвестно';
    if (!acc[n]) acc[n] = { qty: 0, revenue: 0 };
    acc[n].qty += i.quantity || 0;
    acc[n].revenue += (i.price || 0) * (i.quantity || 0);
    return acc;
  }, {});

  const productsText = Object.entries(products)
    .sort((a, b) => b[1].revenue - a[1].revenue)
    .map(([n, d]) => `  - ${n}: ${d.qty} шт., выручка ${fmt(d.revenue)} KZT`)
    .join('\n');

  return `Данные магазина Nova (корректирующее бельё, Казахстан):

Общая статистика:
  - Заказов: ${count}
  - Общая сумма: ${fmt(totalSum)} KZT
  - Средний чек: ${fmt(avg)} KZT

Города:
${citiesText}

Источники трафика (UTM):
${utmsText}

Топ товаров:
${productsText}`;
}

export default async function handler(req, res) {
  if (req.method !== 'POST') {
    return res.status(405).json({ error: 'Method not allowed' });
  }

  if (!OPENAI_API_KEY || !SUPABASE_URL || !SUPABASE_SERVICE_KEY) {
    return res.status(500).json({ error: 'Не заданы переменные окружения на сервере' });
  }

  try {
    // Загрузить данные из Supabase
    const [orders, items] = await Promise.all([
      supabaseFetch('orders?select=total,city,utm_source'),
      supabaseFetch('order_items?select=product_name,quantity,price'),
    ]);

    const summary = buildSummary(orders, items);

    // Вызвать OpenAI
    const aiRes = await fetch('https://api.openai.com/v1/chat/completions', {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${OPENAI_API_KEY}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        model: 'gpt-4o-mini',
        temperature: 0.7,
        messages: [
          { role: 'system', content: SYSTEM_PROMPT },
          { role: 'user', content: summary },
        ],
      }),
    });

    if (!aiRes.ok) {
      const err = await aiRes.json().catch(() => ({}));
      return res.status(502).json({ error: err.error?.message || 'Ошибка OpenAI' });
    }

    const aiJson = await aiRes.json();
    const lines = aiJson.choices[0].message.content
      .trim()
      .split('\n')
      .filter((l) => l.trim());

    // Обновить таблицу insights
    await supabaseFetch('insights?id=neq.0', { method: 'DELETE' });
    await supabaseFetch('insights', {
      method: 'POST',
      body: JSON.stringify(lines.map((content) => ({ content }))),
    });

    // Вернуть свежие инсайты
    const fresh = await supabaseFetch(
      'insights?select=*&order=generated_at.desc'
    );
    return res.status(200).json({ insights: fresh });

  } catch (err) {
    return res.status(500).json({ error: err.message });
  }
}
